[lldb] Fix FindProcessImpl() for iOS simulators

royitaqi opened the pull request:
# Benefit

This patch fixes:
1. After `platform select ios-simulator`, `platform process list` will now print processes which are running in the iOS simulator. Previously, no process will be listed.
2. After `platform select ios-simulator`, `platform attach --name <name>` will succeed. Previously, it will error out saying no process is found.


# Several bugs that is being fixed

1. During the process listing, add `aarch64` to the list of CPU types for which iOS simulators are checked for.
2. Given a candidate process, when checking for simulators, the original code will find the desired environment variable (`SIMULATOR_UDID`) and set the OS to iOS, but then the immediate next environment variable will set it back to macOS.
3. For processes running on simulator, set the triple's `Environment` to `Simulator`, so that such processes can pass the filtering [in this line](https://fburl.com/8nivnrjx). The original code leave it as the default `UnknownEnvironment`.



# Manual test

**With this patch:**
```
royshi-mac-home ~/public_llvm/build % bin/lldb
(lldb) platform select ios-simulator

(lldb) platform process list
240 matching processes were found on "ios-simulator"

PID    PARENT USER       TRIPLE                         NAME
====== ====== ========== ============================== ============================
40511  28844  royshi     arm64-apple-ios-simulator      FocusPlayground // my toy iOS app running on simulator
... // omit
28844  1      royshi     arm64-apple-ios-simulator      launchd_sim

(lldb) process attach --name FocusPlayground
Process 40511 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = signal SIGSTOP
    frame #0: 0x0000000104e3cb70 libsystem_kernel.dylib`mach_msg2_trap + 8
libsystem_kernel.dylib`mach_msg2_trap:
->  0x104e3cb70 <+8>: ret
... // omit
```

**Without this patch:**
```
$ bin/lldb
(lldb) platform select ios-simulator

(lldb) platform process list
error: no processes were found on the "ios-simulator" platform

(lldb) process attach --name FocusPlayground
error: attach failed: could not find a process named FocusPlayground
```


# Unittest

With the patch, the test can list processes on the iOS simulator:
```
$ bin/llvm-lit -av ../llvm-project/lldb/test/API/macosx/simulator/TestSimulatorPlatform.py
...
PASS: LLDB (/Users/royshi/public_llvm/build/bin/clang-arm64) :: test_ios (TestSimulatorPlatform.TestSimulatorPlatformLaunching)
```


Without the patch, the test cannot list processes on the iOS simulator:
```
$ bin/llvm-lit -av ../llvm-project/lldb/test/API/macosx/simulator/TestSimulatorPlatform.py
...
======================================================================
FAIL: test_ios (TestSimulatorPlatform.TestSimulatorPlatformLaunching)
    Test running an iOS simulator binary
----------------------------------------------------------------------
...
AssertionError: False is not true : Command 'platform process list' did not return successfully
Error output:
error: no processes were found on the "ios-simulator" platform
```

royitaqi committed:
Fix macOS FindProcessImpl()

JDevlieghere commented:
Could we test this in `TestSimulatorPlatform.py`?

royitaqi commented:
> Could we test this in `TestSimulatorPlatform.py`?

@JDevlieghere Thanks for pointer. It seems the tests in that file are all **skipped** because of this bug number: `rdar://76995109`.

> UNSUPPORTED: LLDB (/Users/royshi/public_llvm/build/bin/clang-arm64) :: test_ios (TestSimulatorPlatform.TestSimulatorPlatformLaunching) (skipping unconditionally [rdar://76995109])

Did a bit internet search and couldn't find how to find more info about this bug or why these tests are all skipped. Not sure if I should un-skip them.

What's your advice on my next steps?

JDevlieghere commented:
> > Could we test this in `TestSimulatorPlatform.py`?
> 
> @JDevlieghere Thanks for pointer. It seems the tests in that file are all **skipped** because of this bug number: `rdar://76995109`.
> 
> E.g.
> 
> ```
> UNSUPPORTED: LLDB (/Users/royshi/public_llvm/build/bin/clang-arm64) :: test_ios (TestSimulatorPlatform.TestSimulatorPlatformLaunching) (skipping unconditionally [rdar://76995109])
> ```
> 
> Did a bit internet search and couldn't find how to find more info about this bug or why these tests are all skipped. Not sure if I should un-skip them.

Ha, that's my radar, and it's no longer relevant. I bisected an issue with the test suite to that particular test, but the last comment says that it wasn't the culprit after all, so there's no reason it should still be disabled.

royitaqi commented:
> > > Could we test this in `TestSimulatorPlatform.py`?
> > 
> > @JDevlieghere Thanks for pointer. It seems the tests in that file are all **skipped** because of this bug number: `rdar://76995109`.
> > E.g.
> > ```
> > UNSUPPORTED: LLDB (/Users/royshi/public_llvm/build/bin/clang-arm64) :: test_ios (TestSimulatorPlatform.TestSimulatorPlatformLaunching) (skipping unconditionally [rdar://76995109])
> > ```
> > 
> > Did a bit internet search and couldn't find how to find more info about this bug or why these tests are all skipped. Not sure if I should un-skip them.
> 
> Ha, that's my radar, and it's no longer relevant. I bisected an issue with the test suite to that particular test, but the last comment says that it wasn't the culprit after all, so there's no reason it should still be disabled.

Hi @JDevlieghere,

I don't think the test file was very happy when I removed that skip statement (`@skipIf(bugnumber="rdar://76995109")`). 4 tests coudln't build and 1 had assertion error about a mach-o load command (the other 7 skipped for other reasons). See full result: https://pastebin.com/Ka7KKdkC.

When you have the time, I wonder if you could kindly try on your side and see if they work for you.

FWIW I'm using the following command to build and test:
```
ls ~/public_llvm/llvm-project
cd ~/public_llvm/build
cmake -G Ninja -DLLVM_ENABLE_PROJECTS="clang;lldb" -DCMAKE_BUILD_TYPE=Debug -DLLVM_ENABLE_RUNTIMES="libcxx;libcxxabi;libunwind" ~/public_llvm/llvm-project/llvm
ninja lldb-api-test-deps
bin/llvm-lit -sv ../llvm-project/lldb/test/API/macosx/simulator/TestSimulatorPlatform.py
```

royitaqi committed:
Merge branch 'llvm:main' into fix-ios-simulator-process-list-and-attach

royitaqi committed:
Add test

royitaqi commented:
Hi @JDevlieghere ,

Thank you very much for your #142244. I have rebased this patch on top of that and added assertions into the iOS test case to verify that we can get a list of processes on the iOS simulator.

There is something that I don't think blocks this PR but would like to understand:
1. It seems to me the test binary `a.out` isn't running on the simulator (doesn't appear in `platform process list`).
2. My original plan is to assert that we can find `a.out` in the process list. Because of the above, I make the assertion looser, that it verifies that we can get any process list at all. IMHO it still verifies the patch (that we can now get process list, previously completely nothing), but if we can fix 1 then my assertion can be to find `a.out`.

Thanks,
Roy

JDevlieghere commented:
> 1. It seems to me the test binary `a.out` isn't running on the simulator (doesn't appear in `platform process list`). Is this expected/intentional?

I can definitely see the binary running:

```
jonas            46758   0.0  0.0 410596000  22720 s000  SX    2:43PM   0:00.09 /Users/jonas/llvm/build-ra/lldb-test-build.noindex/macosx/simulator/TestSimulatorPlatform.test_ios/a.out
```

If it's not showing up in the `platform process list` then maybe there's a bug in the implementation?

royitaqi committed:
Fix format

royitaqi commented:
@JDevlieghere ,

> I can definitely see the binary running:

Oh interesting. Then I might have got it wrong.

How did you confirm that this is running *on the simulator*? (because `platform process list` would only get those)

I do agree it's running somewhere (because the rest of the test passed). But thought maybe it's running on the host (macOS) machine (thought catalyst can do that but I haven't tried it to know for sure).

royitaqi commented:
Hi @JDevlieghere,

I think I know why: In the `Host::FindProcessImpl()` for iOS simulator, after getting all processes in the macOS, it filters out the processes which are being debugged ([here](https://fburl.com/eid5y2oa)). `a.out` is of course being debugged (by the test case) and got filtered out.   IMHO this is more like a limitation of the test, not a bug in the implementation.

Now, there are a few ways we can go:
1. Accept the patch as is, since we have confirmed that it's not a bug. Downside is that the test isn't super convincing.
2. If we want to make the api test to find `a.out`, we can add an additional parameter to `platform process list` (e.g. `--all-processes`) to skip such filter, then forward such parameter all the way into `Host::FindProcessImpl()`.  LMK whether or not you think this would be useful besides the api test (e.g. would a user want this).
3. Instead of finding `a.out`, we can find other processes which is running on the iOS simulator because of the test. My difficult is that I don't know any of such processes.

Looking forward to your thoughts.


Best,
Roy

royitaqi committed:
Minor updates to code and comments

clayborg commented on lldb/source/Host/macosx/objcxx/Host.mm:
```diff
@@ -621,22 +623,25 @@ DataExtractor data(arg_data.GetBytes(), arg_data_size,
 
           Environment &proc_env = process_info.GetEnvironment();
+          bool is_simulator = false;
           while ((cstr = data.GetCStr(&offset))) {
```
Lets use llvm::StringRef here, you will see why below:
```
while (true) {
  llvm::StringRef env_var(data.GetCStr(&offset));
  if (env_var.empty())
    break;
```

royitaqi replied:
Updated the PR, but not exactly as you suggested (I tried a `for` loop, not a `while`). LMK if that's better or worse.

clayborg commented on lldb/source/Host/macosx/objcxx/Host.mm:
```diff
@@ -621,22 +623,25 @@ DataExtractor data(arg_data.GetBytes(), arg_data_size,
-                process_info.GetArchitecture().GetTriple().setOS(
-                    llvm::Triple::MacOSX);
-            }
+            if (check_for_ios_simulator &&
+                strncmp(cstr, "SIMULATOR_UDID=", strlen("SIMULATOR_UDID=")) ==
+                    0)
```
```
if (check_for_ios_simulator && env_var.starts_with("SIMULATOR_UDID="))
```

royitaqi committed:
A small cleanup

royitaqi commented:
Hi @JDevlieghere, when you have the time, could you kindly review / see my rely to your previous comment?
https://github.com/llvm/llvm-project/pull/139174#issuecomment-2978937345

JDevlieghere commented:
> I think I know why - it's by design: In the `Host::FindProcessImpl()` for iOS simulator, after getting all processes in the macOS, it filters out the processes which are being debugged ([here](https://github.com/llvm/llvm-project/blob/b5dbf8210a57b986b9802304745f4c5c108cf37b/lldb/source/Host/macosx/objcxx/Host.mm#L724)). `a.out` is of course being debugged (by the test case) and got filtered out.

Ah, that's unfortunate, but good catch. In `TestAppleSimulatorOSType` there's a snippet that launched a binary with `simctl`. Could we use that to launch the binary without it being traced?

```
$ xcrun simctl spawn -s <device uuid> <path/to/binary>
```

royitaqi committed:
Run a.out in simulator and find it

royitaqi commented on lldb/test/API/macosx/simulator/hello.cpp:
```diff
@@ -0,0 +1,24 @@
+
+static void print_pid() {
+  #if defined(_WIN32)
+    fprintf(stderr, "PID: %d\n", ::GetCurrentProcessId());
```
I copied this snippet from from `TestAppleSimulatorOSType`'s `main.cpp` (see [code](https://github.com/llvm/llvm-project/blob/c455f4a32d91436c131a751fc9587d7fa3ded614/lldb/test/API/tools/lldb-server/main.cpp#L45-L51)).

However, it's OS-dependent, and I don't have a Windows machine to verify.

Questions:
1. What header files do I need to include for `::GetCurrentProcessId()`?
2. Does the test run in Windows? If not, maybe I can just remove this `#if`.

royitaqi reviewed (commented):
Hi @JDevlieghere, 

> In TestAppleSimulatorOSType there's a snippet that launched a binary with simctl. Could we use that to launch the binary without it being traced?

Done. Kindly take a look when you have the time.

TL;DR changes:
1. Found the snippet you were referring to in `TestAppleSimulatorOSType`.
2. Refactored two functions into `lldbutil`.
3. Used those two functions in `TestSimulatorPlatform` to launch `a.out` separately and find it in the output of `platform process list` (matching both pid and name).
4. In order to do 3, I had to rename `hello.c` and add code so that it prints the pid and also wait for 10s.

I have a **question** regarding 4, about how to make sure that the `hello.cpp` builds in Windows. See inline comment.

royitaqi commented:
Hi @JDevlieghere and @clayborg ,

Could you kindly review when you have the chance? I think the PR is close to finish, except the "how to make sure hello.cpp builds in Windows" question.

TL;DR updates:
1. Called `xcrun simctl spawn -s <device uuid> <path/to/binary>` to launch `a.out` and verified that its name and pid appear in `platform process list`.
2. Did the small fix which @clayborg pointed out, but with a `for` loop (not a `while`).

Thanks,
Roy

JDevlieghere commented:
> I think the PR is close to finish, except the "how to make sure hello.cpp builds in Windows" question.

Does this test need to build on Windows? That test should only run on Darwin, so if it's properly skipped, we shouldn't even try to build the inferior. The Windows-conditional parts will be dead code.

Besides that this LGTM

royitaqi committed:
Remove windows code from hello.cpp

royitaqi commented:
> The Windows-conditional parts will be dead code.
Hi @JDevlieghere, I have moved the Windows related code. Should be good to go (re-running tests on the side).

FWIW, my bad, I overlooked the `@skipUnlessDarwin` in the test.

clayborg commented on lldb/source/Host/macosx/objcxx/Host.mm:
```diff
@@ -621,21 +623,20 @@ DataExtractor data(arg_data.GetBytes(), arg_data_size,
-
-            proc_env.insert(cstr);
+          bool is_simulator = false;
+          for (llvm::StringRef env_var;
+               !(env_var = data.GetCStr(&offset)).empty();) {
```
This might be more clear using a while statement:
```
llvm::StringRef env_var;
while (!(env_var = data.GetCStr(&offset)).empty()) {
```

clayborg reviewed (approved):
One possible nit about switching to using a while loop, but looks good.

royitaqi committed:
Minor update for readability
